src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -739,6 +739,67 @@ class StringGenerator : public Generator<std::string> {
     const size_t _alphabetLength;
 };
 
+class DataSetGenerator : public Generator<std::string> {
+public:
+    DataSetGenerator(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _id{generatorArgs.actorId},
+          _path{node["path"].maybe<std::string>().value()},
+          _pathLength{_path.size()} {
+        if (_pathLength <= 0) {
+            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+                "Data Set Generator requieres non-empty path"));
+        }
+    }
+
+protected:
+    DefaultRandom& _rng;
+    ActorId _id;
+    UniqueGenerator<int64_t> _lengthGen;
+    std::string _path;
+    const size_t _pathLength;
+};
+
+/** `{^ChooseFromDataset:{...}` */
+class RandomDataFromDataset : public DataSetGenerator {
+
+public:
+    RandomDataFromDataset(const Node& node, GeneratorArgs generatorArgs)
+        : DataSetGenerator(node, generatorArgs) {
+            
+            // Open and read the datasets in the constructor
+            // to avoid doing it on each evaluate method call.
+            std::string line;
+            ifs.open(_path, std::ifstream::in);
+            if (ifs.is_open()){
+                while(std::getline(ifs, line))
+                {
+                    // Ignore emtpy lines
+                    if (line.size()!=0){
+                        dataSetContent.push_back(line);
+                    }
+                }
+            }
+            else {
+                BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+                    "The specified file for RandomDataFromDataset cannot be opened or it does not exist"));
+                }
+
+            if (dataSetContent.size() == 0) {
+                BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+                    "The specified file for RandomDataFromDataset is empty"));
+                }
+        }
+
+    std::string evaluate() override {   
+        auto distribution = boost::random::uniform_int_distribution<size_t>{0, dataSetContent.size() - 1};
+        return dataSetContent[distribution(_rng)];
+    }
+private:
+    std::ifstream ifs;
+    std::vector<std::string> dataSetContent;
```

Comment:
+1 on static map somehow. There are a couple of levels of this. 
At its simplest form, we should make sure that each instantiation within an actor shares state. That likely can be done with a static member (maybe?)
Then there is possible sharing between different actors or different parts of the same actor (e.g, multiple operations). I'm not sure the best way to do that.

Reply:
Right now, every time the generator is called and the object is created the file is read into memory. Once the generator has finished, the object is destroyed and memory released. This means that if a particular genny workload has two or more ^ChooseFromDataset, data in memory will be loaded once per generator. I don't know if this is going to be in parallel, since I don't really know if the document generation is multithreaded.

I will prototype something quick when I have time to use a static map, to avoid re-reading the file into memory multiple times.

Reply:
Yes the evaluate method could be called from multiple threads here. For example, if we have 32 threads set on the loader actor, there will be 32 copies of the actor working in parallel to insert documents into a collection, and each will be generating documents. This means that if we have static state to save memory, we'll have to handle thread-safety. One option is a static unordered_map and a static mutex in the generator class.

Reply:
Thank you for your comments, this is done in the last commit.

- Using a static unordered_map to store the file content, where the key is the file path.
- If the key exists, we don't open the file again.
- If the key doesn't exist, we enter a section protected by a mutex that opens the file and stores the data as a vector of strings for that particular key (path).